File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
void helper(int open , int close , string str , int n, vector<string>& ans){
   if(open > n || close > open) return;
   if(open == n && close == n){
      ans.push_back(str);
      return;
   }

   if(open < n){
     str.push_back('(');
     helper(open+1,close, str,n, ans);
     str.pop_back();
   }

   if(close < open){
    str.push_back(')');
     helper(open , close+1,str,n,ans);
     str.pop_back();
   }


}
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        string str;
        helper(0,0,str,n,ans);
        return ans;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public void helper(List<String>ans,int left,int right,int n,String s){
        if(left == n && right ==n) ans.add(s);
     if(left < n){
         helper(ans,left+1,right,n,s+"(");
     }
     if(left > right){
       helper(ans,left,right+1,n,s+")");
     }
    }
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        helper(ans,0,0,n,"");
        return ans;
    }
}